### PR TITLE
Adding support for Sarus container engine

### PR DESCRIPTION
**Pull request by marcodelapierre**
Hi @pditommaso @evanfloden ,

I hope this feature can be useful for part of the community.
Sarus is a container engine developed and in use at CSCS in Switzerland, see https://sarus.readthedocs.io.

I have successfully run `make test` locally with this branch.
I have also successfully tested the branch in action with the `nextflow-io/hello` workflow (both plain workflow, and some modifications, to test all options for the new feature).

@madeeks , @lichinka ,
pinging you, so that you are aware of this new interface to Sarus being proposed for the Nextflow workflow engine, and also in case you have any feedback for this PR.

### CODE REVIEW

**Comment by marcodelapierre** [comment]
note: the commit list looks a bit weird to me, compared to PRs I usually open. 
I had to do a sign-off rebase due to missing commit signatures, which is what caused the mixing between my commits and recent commits by others.

**Comment by pditommaso** [comment]
You are a star!

File: modules/nextflow/src/main/groovy/nextflow/container/SarusBuilder.groovy [hunk 1]
```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright 2020-2022, Seqera Labs
+ * Copyright 2013-2019, Centre for Genomic Regulation (CRG)
```
Comment (pditommaso) [inline modules/nextflow/src/main/groovy/nextflow/container/SarusBuilder.groovy:3]: I think Pawsay should be reported instead of CRG 😉

**Comment by pditommaso** [comment]
Can Sarun pull the image on-the-fly as it can be done with `docker pull` or it does need to pulled ahead of the run?

File: modules/nextflow/src/main/groovy/nextflow/container/SarusBuilder.groovy (hunk 1 above)
Comment (marcodelapierre) [reply modules/nextflow/src/main/groovy/nextflow/container/SarusBuilder.groovy:3]: done!

File: modules/nextflow/src/main/groovy/nextflow/executor/BashWrapperBuilder.groovy [hunk 2]
```diff
@@ -141,7 +141,7 @@ class BashWrapperBuilder {
     }
 
     protected boolean fixOwnership() {
-        systemOsName == 'Linux' && containerConfig?.fixOwnership && runWithContainer && containerConfig.engine == 'docker' // <-- note: only for docker (shifter is not affected)
+        systemOsName == 'Linux' && containerConfig?.fixOwnership && runWithContainer && containerConfig.engine == 'docker' // <-- note: only for docker (sarus and shifter are not affected)
```
[the remark this answers is not in the document]
Comment (marcodelapierre) [reply modules/nextflow/src/main/groovy/nextflow/executor/BashWrapperBuilder.groovy:144]: done!

File: docs/container.rst [hunk 3]
```diff
@@ -315,6 +315,65 @@ Advanced settings
...
+from `Docker Hub <https://hub.docker.com/>`_ and then submit jobs which run entirely within the container.
+
+Prerequisites
+-------------
+
+You need Sarus installed in your execution environment,
+i.e: your personal computer or a distributed cluster, depending
+on where you want to run your pipeline.
+
+.. note:: This feature requires Sarus version 1.5.1 (or later) and Nextflow 23.04.0 (or later).
```
Comment (pditommaso) [inline docs/container.rst:335]:
```suggestion
.. note:: This feature requires Sarus version 1.5.1 (or later) and Nextflow 22.12.0-edge (or later).
```

**Comment by pditommaso** [comment]
If the automatic pull then this mention in the docs looks inconsistent to me or am I wrong? 

https://github.com/nextflow-io/nextflow/pull/3470/files#diff-22c9ee42ec9ef1d2237cadf42e6d24114e3f0fd8d50b98299fa1f13fcbb76853R351

**Comment by Madeeks** [comment]
> Can Sarun pull the image on-the-fly as it can be done with docker pull or it does need to bepulled ahead of the run?

> Unfortunately, at the moment automatic pull is not support in Sarus (see [eth-cscs/sarus#6](https://github.com/eth-cscs/sarus/issues/6))

Hi @marcodelapierre @pditommaso, thanks for your interest in Sarus and for this development!
I just wanted to make some quick comments.

One of the motivations for not including an automatic image pull in `sarus run` is to avoid multiple concurrent pulls if a user launches a large scale job without having the image available. This is not a technical limitation, Sarus has atomic update mechanisms for images and repository metadata which work quite well, but it's mainly a way to prevent time/resource waste and encourage a "pull once, deploy at scale" usage model, for which Sarus was originally conceived.

Since Sarus 1.5.0, `sarus pull` returns immediately if the requested image is already available locally, so prepending a pull before a run should have negligible wall clock time impact after the first invocation. Apologies for not mentioning this in the related issue.

**Comment by pditommaso** [comment]
Hi @Madeeks thanks for the heads up! this is indeed very interesting. 

I think adding the pull before the runs would make the experience more consistent with the expectation of nextflow users. What do you think @marcodelapierre?

**Comment by marcodelapierre** [comment]
Thanks @Madeeks  and @pditommaso for your feedback! :-) 

Indeed, I am already pre-pulling images to be consistent with the Nextflow user experience: https://github.com/marcodelapierre/nextflow/blob/add/sarus_containers/modules/nextflow/src/main/groovy/nextflow/container/SarusBuilder.groovy#L84-L92

**Comment by pditommaso** [comment]
Works like a charm! 

Screenshot 2022-12-06 at 16 22 09

**Pull request merged**